addons/xivbar/events.lua

```diff
@@ -0,0 +1,92 @@
+--[[
+        Copyright © 2017, SirEdeonX
+        All rights reserved.
+
+        Redistribution and use in source and binary forms, with or without
+        modification, are permitted provided that the following conditions are met:
+
+            * Redistributions of source code must retain the above copyright
+              notice, this list of conditions and the following disclaimer.
+            * Redistributions in binary form must reproduce the above copyright
+              notice, this list of conditions and the following disclaimer in the
+              documentation and/or other materials provided with the distribution.
+            * Neither the name of xivbar nor the
+              names of its contributors may be used to endorse or promote products
+              derived from this software without specific prior written permission.
+
+        THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+        ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+        WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+        DISCLAIMED. IN NO EVENT SHALL SirEdeonX BE LIABLE FOR ANY
+        DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+        (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+        LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+        ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+        (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+        SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+]]
+
+-- Bind Events
+-- ON LOAD
+windower.register_event('load',function()
+    if windower.ffxi.get_info().logged_in then
+        initialize()
+    end
+end)
+
+-- ON LOGIN
+windower.register_event('login',function()
+    initialize()
+end)
+
+-- ON LOGOUT
+windower.register_event('logout',function()
+    hide()
+end)
+
+-- BIND EVENTS
+windower.register_event('hp change', function(new, new)
+    hp_update = true
+end)
+
+windower.register_event('hpmax change', function(new, old)
+    hp_update = true
+end)
+
+windower.register_event('mp change', function(new, old)
+    mp_update = true
+end)
+
+windower.register_event('mpmax change', function(new, old)
+    mp_update = true
+end)
+
+windower.register_event('tp change', function(new, old)
+    tp_update = true
+end)
+
+windower.register_event('prerender', function()
```

Comment:
Just to add to that, `get_party` scales with party members (as was observed), but generally mostly pushed the mob array entry for each party member, so it's basically `<some overhead> + get_mob_by_index * <number of party members>`. `get_player` on the other hand pushes a lot that's not immediately obvious, like all merits, all job points (which are each many tables with subentries), buffs, skills, vitals, jobs, etc., all of which add up to quite a bit. Still, as is obvious from the timing, `get_items` trumps them all, which is why I added the specific overloads for specific bags/slots as well as the `add item` and `remove item` events, to not have to constantly query all bags to monitor changes.

As a side note, the performance of `get_mob_by_id` depends on the type of the ID. If it's an NPC ID with a non-zero first byte, it just falls back on `get_mob_by_index`. If it's a player ID it iterates over the player index range and compares IDs. However, empty zones are not necessarily faster in that regard, as indices cycle, so it's just a matter of luck.